### DISCUSSION TITLE
Question regarding implementation of exceptions in cpp2.

### DISCUSSION THREAD

**Discussion by AvangardAA** [hsutter/cppfront#1149] [Q&A] [answered]
Greetings!
Currently wondering where its possible to find a docs or examples on semantics of try {} catch {} and throw in cpp2. I've seen issues #111 and #361 but there is no explicit answer on how to define a try-catch block or how to catch what's thrown by 'throws' word i've seen in listed issues.
Thanks in advance.

**Comment by DyXel** [comment, marked as answer]
Exceptions haven't yet been implemented in Cpp2/cppfront. AFAIK, there's plan for them in the future.

**Comment by MatthieuHernandez** [comment]
If I import a cpp1 library into my cpp2 project, exceptions from my library just result in abort() and I can't retrieve the exception message.
Really annoying :weary:

**Reply by JohelEGP** [reply]
Do you mean contract assertions?
Replace the default handler.
See <https://hsutter.github.io/cppfront/cpp2/contracts/>.

**Reply by MatthieuHernandez** [reply]
Yes, it seems so, but the last example on the page doesn't compile (I'm in version 0.7.4).
```
Error: static assertion failed due to requirement 'cpp2::impl::dependent_false<cpp2::contract_group &>::value': this function call syntax tries 'obj.func(...)', then 'func(obj,...);', but both failed - if this function call is passing a local variable that will be modified by the function, but that variable is never used again in the function so the new value is never used, that's likely the problem - if that's what you intended, add another line '_ = obj;' afterward to explicitly discard the new value of the object
   16 |     static_cast<void>(CPP2_UFCS(set_handler)(cpp2::cpp2_default, &call_my_framework));
```

**Reply by JohelEGP** [reply]
See #1334.

**Reply by MatthieuHernandez** [reply]
Sorry, I don't understand what this has to do with my problem, without these lines:
```cpp
call_my_framework:(msg: * const char) = {
    std::cout << msg << std::endl;
}

main:() -> int = {
    cpp2::default.set_handler(call_my_framework&);
```
my code compiles correctly.

**Reply by JohelEGP** [reply]
You're right, I gave the wrong diagnostic.
Now I see that it was UFCS that failed to find a candidate.
Maybe the signature to set the handler changed.

**Reply by MatthieuHernandez** [reply]
Yes, it works if I use cpp1 to define the correct signature.
```cpp
void call_my_framework(const char * msg, std::source_location) {
    std::cout << "msg" << std::endl;
    exit(0);
}

main:() -> int = {
    cpp2::default.set_handler(call_my_framework&);
    assert<default>(false, "this is a test, this is only a test");
```
But that's not at all what I wanted to do in the first place.
In this example:
```cpp
#include "player.h2"
#include <snn/neural_network/StraightforwardNeuralNetwork.hpp>

createAi:() = {
    layers: std::vector<snn::LayerModel> = ();
    neuralNetwork:= snn::StraightforwardNeuralNetwork(layers);
    neuralNetwork.saveAs("./snn_models/first_model.snn");
}
```
I'd like my cpp2 code to correctly handle exceptions coming from my external library `snn`.

**Reply by JohelEGP** [reply]
Since you're writing a library, you don't want to replace the default handler.
IIRC the way to have a contract use a specific handler is
by passing it as the first template argument,
e.g. `assert<my_handler>(a && b);`.

**Reply by MatthieuHernandez** [reply]
```cpp
void test() {
    throw std::runtime_error("message to retrieve");
}

main:() -> int = {
    test();
    return 0;
}
```
I can't figure out what to write in the main function to get my error message displayed at runtime.
(Assuming that the test method comes from an imported cpp1 library.)

**Reply by JohelEGP** [reply]
Sounds like you need `try`/`catch` syntax in Cpp2, which isn't implemented yet.

**Reply by MatthieuHernandez** [reply]
Ah, I have the impression that this is a CMake and compiler option related problem more than a cpp2 problem. Sorry for the inconvenience and thanks for the help. I'll keep working on a solution on my own.

**Reply by MatthieuHernandez** [reply]
I've found the perfect simple solution, just encapsulate all the code in a try catch 😊 
```cpp
cpp2_main:() -> int = {
    // My entire cpp2 code.
}

int main() {
    try {  // CPP2 workaround: Try catch not yet supported.
        return cpp2_main();
    } catch (const std::exception& e) {
        std::cerr << "Caught exception: " << e.what() << std::endl;
    }
    return 0;
}
```
